src/unity/toolkits/evaluation/metrics.cpp

```diff
@@ -25,6 +25,75 @@
 namespace turi {
 namespace evaluation {
 
+namespace {
+
+std::vector<std::string> get_default_classifier_metrics() {
+  return {"accuracy", "auc", "precision", "recall", "f1_score", "log_loss",
+          "confusion_matrix", "roc_curve"};
+}
+
+std::vector<std::string> get_classifier_metrics(const std::string& metric) {
+  // Validate metric and determine list of metrics to compute.
+  std::vector<std::string> metrics = get_default_classifier_metrics();
+  if (metric != "auto") {
+    // If the caller didn't request "auto", then adjust the list of metrics.
+    if (metric == "report") {
+      // Add the per-class report to the standard list of metrics.
+      metrics.push_back("report_by_class");
+    } else {
+      // Just compute the requested metric, if valid.
+      if (std::find(metrics.begin(), metrics.end(), metric) == metrics.end()) {
+        log_and_throw("Unsupported metric " + metric);
+      } else {
+        metrics = {metric};
+      }
+    }
+  }
+  return metrics;
+}
+
+gl_sarray get_prediction_probability_vectors(const gl_sarray& predictions,
+                                             const flex_list& class_labels) {
+  // Canonicalize predictions into flex_vec.
+  gl_sarray result;
+  switch (predictions.dtype()) {
+  case flex_type_enum::VECTOR:
+    result = predictions;
+    break;
+  case flex_type_enum::ND_VECTOR:
+    result = predictions.astype(flex_type_enum::VECTOR);
+    break;
+  case flex_type_enum::DICT: {
+    // Require/assume the dictionary to map class labels to probabilities.
+
+    // Build a fast lookup table from label to flex_vec index.
+    auto class_to_index =
+        std::make_shared<std::unordered_map<flexible_type, size_t>>();
+    for (size_t i = 0; i < class_labels.size(); ++i) {
+      class_to_index->emplace(class_labels[i], i);
+    }
+
+    // Use the lookup table to write a flex_vec for each flex_dict.
+    auto flatten = [class_to_index](const flexible_type& ft) {
+      const flex_dict& dict = ft.get<flex_dict>();
+      flex_vec vec(dict.size());
+      for (const auto& kv : dict) {
+        vec[class_to_index->at(kv.first)] = kv.second;
+      }
+      return vec;
+    };
+    result = predictions.apply(flatten, flex_type_enum::VECTOR);
+    break;
+  }
+  default:
+    log_and_throw("Could not convert predictions to probability vectors for "
```

Comment:
I think we're missing a `+` at the end of this line?

Reply:
Actually, I don't believe `+` is defined for C string literals. However, if the compiler sees two C string literals separated only by whitespace, it will simply concatenate them.

See also: the third paragraph under "Notes" at https://en.cppreference.com/w/cpp/language/string_literal

Reply:
Got it, thanks!